src/mds/MDSRank.cc

```diff
@@ -1019,14 +1019,7 @@ void MDSRank::ProgressThread::shutdown()
 
 bool MDSRankDispatcher::ms_dispatch(const cref_t<Message> &m)
 {
-  if (m->get_source().is_mds()) {
```

Comment:
Yeah, the reply will be client type. And here it will crash the MDS.

Reply:
Will it never be a MDS? Then it won't hit this `if` block, right? So how will it crash the MDS?

Reply:
The reintegrate will send a client type rename request to the MDS and the reply of the rename will also be a client type. And currently the MDS do not allow to receive a client type request or reply from other MDSs.

And then if will crash in the following code:

`ceph_assert(op);`

Reply:
So if this request was hitting this `if (m->get_source().is_mds()) {` for a request of type client but from a MDS, do you think `if (m->get_source().is_client()) {` will recognize this? I mean it recognized it as MDS and that's why it hit that if condition right?

Reply:
It won't. I just removed the `if (m->get_source().is_mds()) {}` code to avoid the MDS assert in this case.

The `if (m->get_source().is_client()) {}` will do what it should do as before to handle the msgs from clients, none business with the msgs from MDSs.

Reply:
So you mean MDS is not welcome in this part of code, if so then if we ever have a MDS sending requests(client type) which it is sending(cause we discussed the assert failure), I think it's better to log, isn't it? or warn? What do you think?

Reply:
Another questions: Can't we stop MDS  from sending such requests here? or this is intentional?

Reply:
> So you mean MDS is not welcome in this part of code, if so then if we ever have a MDS sending requests(client type) which it is sending(cause we discussed the assert failure), I think it's better to log, isn't it? or warn? What do you think?

Currently the MDS will send a client type request `_OP_RENAME` to other MDSs to do the dentry reintegrate. Though this is also a client type op sent from a MDS, while the `MClientRequest` is an `MMDSOp` type and the `op` will be not `nullptr`, so the `ceph_assert(op)` won't be triggered in this if scope.

But the `MClientReply` is not an `MMDSOp` type, and then `ceph_assert(op)` will be triggered if it comes from an MDS.

> Another questions: Can't we stop MDS from sending such requests here? or this is intentional?

Yes we can, but this is just a normal reply of a client request and if we add a shadow internal OP for the normal `MClientReply` it will introduce complex for the code, which IMO it doesn't make much sense.

Reply:
Or maybe we could fix this by:

```
if (m->get_source().is_mds()) {
    const Message *msg = m.get();
    const MMDSOp *op = dynamic_cast<const MMDSOp*>(msg);
+   const MMDSOp *op1 = dynamic_cast<const SafeMessage*>(msg);
    if (!op && !op1)
-      dout(0) << typeid(*msg).name() << " is not an MMDSOp type" << dendl;
-   ceph_assert(op);
+     dout(0) << typeid(*msg).name() << " is not an MMDSOp or SafeMessage type" << dendl;
+   ceph_assert(op || op1);
  }
  else if (m->get_source().is_client()) {
```
instead of deleting it.